nevergrad/parametrization/data.py

```diff
@@ -0,0 +1,353 @@
+        if self._value.shape != value.shape:
+            raise ValueError(f"Cannot set array of shape {self._value.shape} with value of shape {value.shape}")
+        if not BoundChecker(*self.bounds)(self.value):
+            raise ValueError("New value does not comply with bounds")
+        if self.exponent is not None and np.min(value.ravel()) <= 0:
+            raise ValueError("Logirithmic values cannot be negative")
+        self._value = value
+
+    def sample(self: A) -> A:
+        if not self.full_range_sampling:
+            return super().sample()
+        child = self.spawn_child()
+        std_bounds = tuple(self._to_std_space(b) for b in self.bounds)  # type: ignore
+        diff = std_bounds[1] - std_bounds[0]
+        child.set_standardized_data(std_bounds[0] + np.random.uniform(0, 1, size=diff.shape) * diff, deterministic=False)
+        return child
+
+    def set_bounds(self: A, a_min: BoundValue = None, a_max: BoundValue = None,
+                   method: str = "clipping", full_range_sampling: bool = False) -> A:
+        """Bounds all real values into [a_min, a_max] using a provided method
+
+        Parameters
+        ----------
+        a_min: float or None
+            minimum value
+        a_max: float or None
+            maximum value
+        method: str
+            "clipping", "constraint", "tanh" or "arctan"
+        full_range_sampling: bool
+            whether calling the "sample" method of the parameter should sample uniformly (or log-uniformly) on the whole
+            range of the bounds instead of sampling using a mutation on the current value
+
+        Notes
+        -----
+        - "tanh" reaches the boundaries really quickly, while "arctan" is much softer
+        - only "clipping" accepts partial bounds (None values)
+        """  # TODO improve description of methods
+        bounds = tuple(a if isinstance(a, np.ndarray) or a is None else np.array([a], dtype=float) for a in (a_min, a_max))
+        both_bounds = all(b is not None for b in bounds)
+        # preliminary checks
+        if self.bound_transform is not None:
+            raise RuntimeError("A bounding method has already been set")
+        if full_range_sampling and not both_bounds:
+            raise ValueError("Cannot use full range sampling if both bounds are not set")
+        checker = BoundChecker(*bounds)
+        if not checker(self.value):
+            raise ValueError("Current value is not within bounds, please update it first")
+        if not (a_min is None or a_max is None):
+            if (bounds[0] >= bounds[1]).any():  # type: ignore
+                raise ValueError(f"Lower bounds {a_min} should be strictly smaller than upper bounds {a_max}")
+        # update instance
+        transforms = dict(clipping=trans.Clipping, arctan=trans.ArctanBound, tanh=trans.TanhBound)
+        if method in transforms:
+            if self.exponent is not None and method != "clipping":
```

Comment:
nit: replace the hardcoded strings with constants, same for `self._parameters["sigma"]` (maybe when you are done with experimenting :) )

Reply:
What do you mean? I don't understand it :s

Reply:
```suggestion
            if self.exponent is not None and method != type(self).CLIPPING:
```

And also `self._parameters["sigma"]` -> `self._parameters[type(self).SIGMA]`

A bit longer to write, but allows to catch errors / typos